### PR TITLE
Expose stream parameter in public strings convert APIs

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Add stream parameter to public APIs:

- `cudf::strings::to_booleans()`
- `cudf::strings::from_booleans()`
- `cudf::strings::to_timestamps()`
- `cudf::strings::from_timestamps()`
- `cudf::strings::is_timestamp()`
- `cudf::strings::to_durations()`
- `cudf::strings::from_durations()`
- `cudf::strings::to_fixed_point()`
- `cudf::strings::from_fixed_point()`
- `cudf::strings::to_floats()`
- `cudf::strings::from_floats()`
- `cudf::strings::is_float()`
- `cudf::strings::to_integers()`
- `cudf::strings::from_integers()`
- `cudf::strings::is_integer()`
- `cudf::strings::hex_to_integers()`
- `cudf::strings::integers_to_hex()`
- `cudf::strings::is_hex()`
- `cudf::strings::ipv4_to_integers()`
- `cudf::strings::integers_to_ipv4()`
- `cudf::strings::is_ipv4()`
- `cudf::strings::url_encode()`
- `cudf::strings::url_decode()`
- `cudf::strings::format_list_column()`

Also cleaned up some of the doxygen comments and removed some default parameters.

Reference #13744 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/strings/convert/convert_booleans.hpp [hunk 1]
```diff
@@ -53,16 +55,18 @@ std::unique_ptr<column> to_booleans(
...
+ * @param false_string String to use for false in the output column
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New strings column
  */
 std::unique_ptr<column> from_booleans(
   column_view const& booleans,
-  string_scalar const& true_string    = string_scalar("true"),
-  string_scalar const& false_string   = string_scalar("false"),
+  string_scalar const& true_string,
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_booleans.hpp:67]: `booleans` (above) should be changed to `input` too.

File: cpp/include/cudf/strings/convert/convert_datetime.hpp [hunk 2]
```diff
@@ -231,19 +235,21 @@ std::unique_ptr<column> is_timestamp(
...
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New strings column with formatted timestamps
  */
 std::unique_ptr<column> from_timestamps(
   column_view const& timestamps,
   std::string_view format             = "%Y-%m-%dT%H:%M:%SZ",
   strings_column_view const& names    = strings_column_view(column_view{
     data_type{type_id::STRING}, 0, nullptr, nullptr, 0}),
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_datetime.hpp:252]: `timestamps` (above) ==> `input`?

File: cpp/include/cudf/strings/convert/convert_durations.hpp [hunk 3]
```diff
@@ -115,15 +117,17 @@ std::unique_ptr<column> to_durations(
...
- * @return New strings column with formatted durations.
+ *        Default format is ""%D days %H:%M:%S".
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return New strings column with formatted durations
  */
 std::unique_ptr<column> from_durations(
   column_view const& durations,
   std::string_view format             = "%D days %H:%M:%S",
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_durations.hpp:130]: `durations` (above) ==> `input`?

File: cpp/include/cudf/strings/convert/convert_floats.hpp [hunk 4]
```diff
@@ -39,14 +39,16 @@ namespace strings {
...
+ * @param strings Strings instance for this operation
+ * @param output_type Type of float numeric column to return
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New column with floats converted from strings
  */
 std::unique_ptr<column> to_floats(
   strings_column_view const& strings,
   data_type output_type,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_floats.hpp:51]: `strings` (above) ==> `input`?

File: cpp/include/cudf/strings/convert/convert_floats.hpp [hunk 5]
```diff
@@ -62,12 +64,14 @@ std::unique_ptr<column> to_floats(
...
- * @param mr Device memory resource used to allocate the returned column's device memory.
- * @return New strings column with floats as strings.
+ * @param floats Numeric column to convert
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New strings column with floats as strings
  */
 std::unique_ptr<column> from_floats(
   column_view const& floats,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_floats.hpp:74]: `floats` (above) ==> `input`?

File: cpp/include/cudf/strings/convert/convert_integers.hpp [hunk 6]
```diff
@@ -67,12 +69,14 @@ std::unique_ptr<column> to_integers(
...
- * @param mr Device memory resource used to allocate the returned column's device memory.
- * @return New strings column with integers as strings.
+ * @param integers Numeric column to convert
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New strings column with integers as strings
  */
 std::unique_ptr<column> from_integers(
   column_view const& integers,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_integers.hpp:79]: `integers` (above) ==> `input`?

File: cpp/include/cudf/strings/convert/convert_ipv4.hpp [hunk 7]
```diff
@@ -71,12 +73,14 @@ std::unique_ptr<column> ipv4_to_integers(
...
- * @param mr Device memory resource used to allocate the returned column's device memory.
- * @return New strings column.
+ * @param integers Integer (INT64) column to convert
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New strings column
  */
 std::unique_ptr<column> integers_to_ipv4(
   column_view const& integers,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
```
Comment (ttnghia) [inline cpp/include/cudf/strings/convert/convert_ipv4.hpp:83]: `integers` (above) ==> `input`?

File: cpp/src/strings/convert/convert_booleans.cu [hunk 8]
```diff
@@ -156,10 +157,11 @@ std::unique_ptr<column> from_booleans(column_view const& booleans,
 std::unique_ptr<column> from_booleans(column_view const& booleans,
                                       string_scalar const& true_string,
                                       string_scalar const& false_string,
+                                      rmm::cuda_stream_view stream,
                                       rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::from_booleans(booleans, true_string, false_string, cudf::get_default_stream(), mr);
+  return detail::from_booleans(booleans, true_string, false_string, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_booleans.cu:164]:
```suggestion
  return detail::from_booleans(input, true_string, false_string, stream, mr);
```

File: cpp/src/strings/convert/convert_datetime.cu [hunk 9]
```diff
@@ -1168,10 +1170,11 @@ std::unique_ptr<column> from_timestamps(column_view const& timestamps,
 std::unique_ptr<column> from_timestamps(column_view const& timestamps,
                                         std::string_view format,
                                         strings_column_view const& names,
+                                        rmm::cuda_stream_view stream,
                                         rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::from_timestamps(timestamps, format, names, cudf::get_default_stream(), mr);
+  return detail::from_timestamps(timestamps, format, names, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_datetime.cu:1177]:
```suggestion
  return detail::from_timestamps(input, format, names, stream, mr);
```

File: cpp/src/strings/convert/convert_floats.cu [hunk 10]
```diff
@@ -120,10 +120,11 @@ std::unique_ptr<column> to_floats(strings_column_view const& strings,
 
 std::unique_ptr<column> to_floats(strings_column_view const& strings,
                                   data_type output_type,
+                                  rmm::cuda_stream_view stream,
                                   rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::to_floats(strings, output_type, cudf::get_default_stream(), mr);
+  return detail::to_floats(strings, output_type, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_floats.cu:127]:
```suggestion
  return detail::to_floats(input, output_type, stream, mr);
```

File: cpp/src/strings/convert/convert_hex.cu [hunk 11]
```diff
@@ -280,24 +280,27 @@ std::unique_ptr<column> integers_to_hex(column_view const& input,
...
+  return detail::hex_to_integers(strings, output_type, stream, mr);
 }
 
 std::unique_ptr<column> is_hex(strings_column_view const& strings,
+                               rmm::cuda_stream_view stream,
                                rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::is_hex(strings, cudf::get_default_stream(), mr);
+  return detail::is_hex(strings, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_hex.cu:295]:
```suggestion
  return detail::is_hex(input, stream, mr);
```

File: cpp/src/strings/convert/convert_integers.cu [hunk 12]
```diff
@@ -203,18 +203,20 @@ std::unique_ptr<column> is_integer(strings_column_view const& strings,
 
 // external APIs
 std::unique_ptr<column> is_integer(strings_column_view const& strings,
+                                   rmm::cuda_stream_view stream,
                                    rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::is_integer(strings, cudf::get_default_stream(), mr);
+  return detail::is_integer(strings, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_integers.cu:210]:
```suggestion
  return detail::is_integer(input, stream, mr);
```

File: cpp/src/strings/convert/convert_integers.cu [hunk 13]
```diff
@@ -203,18 +203,20 @@ std::unique_ptr<column> is_integer(strings_column_view const& strings,
...
 }
 
 std::unique_ptr<column> is_integer(strings_column_view const& strings,
                                    data_type int_type,
+                                   rmm::cuda_stream_view stream,
                                    rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::is_integer(strings, int_type, cudf::get_default_stream(), mr);
+  return detail::is_integer(strings, int_type, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_integers.cu:219]:
```suggestion
  return detail::is_integer(input, int_type, stream, mr);
```

File: cpp/src/strings/convert/convert_integers.cu [hunk 14]
```diff
@@ -304,10 +306,11 @@ std::unique_ptr<column> to_integers(strings_column_view const& strings,
 // external API
 std::unique_ptr<column> to_integers(strings_column_view const& strings,
                                     data_type output_type,
+                                    rmm::cuda_stream_view stream,
                                     rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::to_integers(strings, output_type, cudf::get_default_stream(), mr);
+  return detail::to_integers(strings, output_type, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_integers.cu:313]:
```suggestion
  return detail::to_integers(input, output_type, stream, mr);
```

File: cpp/src/strings/convert/convert_ipv4.cu [hunk 15]
```diff
@@ -103,10 +103,11 @@ std::unique_ptr<column> ipv4_to_integers(strings_column_view const& strings,
 
 // external API
 std::unique_ptr<column> ipv4_to_integers(strings_column_view const& strings,
+                                         rmm::cuda_stream_view stream,
                                          rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::ipv4_to_integers(strings, cudf::get_default_stream(), mr);
+  return detail::ipv4_to_integers(strings, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_ipv4.cu:110]:
```suggestion
  return detail::ipv4_to_integers(input, stream, mr);
```

File: cpp/src/strings/convert/convert_ipv4.cu [hunk 16]
```diff
@@ -223,17 +224,19 @@ std::unique_ptr<column> is_ipv4(strings_column_view const& strings,
 // external API
 
 std::unique_ptr<column> integers_to_ipv4(column_view const& integers,
+                                         rmm::cuda_stream_view stream,
                                          rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::integers_to_ipv4(integers, cudf::get_default_stream(), mr);
+  return detail::integers_to_ipv4(integers, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_ipv4.cu:231]:
```suggestion
  return detail::integers_to_ipv4(input, stream, mr);
```

File: cpp/src/strings/convert/convert_ipv4.cu [hunk 17]
```diff
@@ -223,17 +224,19 @@ std::unique_ptr<column> is_ipv4(strings_column_view const& strings,
...
 }
 
 std::unique_ptr<column> is_ipv4(strings_column_view const& strings,
+                                rmm::cuda_stream_view stream,
                                 rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::is_ipv4(strings, cudf::get_default_stream(), mr);
+  return detail::is_ipv4(strings, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_ipv4.cu:239]:
```suggestion
  return detail::is_ipv4(input, stream, mr);
```

File: cpp/src/strings/convert/convert_urls.cu [hunk 18]
```diff
@@ -149,10 +149,11 @@ std::unique_ptr<column> url_encode(strings_column_view const& input,
 
 // external API
 std::unique_ptr<column> url_encode(strings_column_view const& strings,
+                                   rmm::cuda_stream_view stream,
                                    rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::url_encode(strings, cudf::get_default_stream(), mr);
+  return detail::url_encode(strings, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_urls.cu:156]:
```suggestion
  return detail::url_encode(input, stream, mr);
```

File: cpp/src/strings/convert/convert_urls.cu [hunk 19]
```diff
@@ -429,10 +430,11 @@ std::unique_ptr<column> url_decode(strings_column_view const& strings,
 // external API
 
 std::unique_ptr<column> url_decode(strings_column_view const& strings,
+                                   rmm::cuda_stream_view stream,
                                    rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  return detail::url_decode(strings, cudf::get_default_stream(), mr);
+  return detail::url_decode(strings, stream, mr);
```
Comment (ttnghia) [inline cpp/src/strings/convert/convert_urls.cu:437]:
```suggestion
  return detail::url_decode(input, stream, mr);
```

**Review by ttnghia — COMMENTED** [review]
Probably there will be a lot more changes needed to change `strings`, `integers`, etc. into `input` name in the internal APIs.

**Comment by davidwendt** [comment]
> Probably there will be a lot more changes needed to change `strings`, `integers`, etc. into `input` name in the internal APIs.

The intention was to change `strings` to `input` because `strings` is part of the namespace name and sometimes produces confusing compile errors. We don't have the same issue with other types like `booleans` and `integers` so I'd like to leave those unchanged.

File: cpp/include/cudf/strings/convert/convert_durations.hpp [hunk 20]
```diff
@@ -115,15 +117,17 @@ std::unique_ptr<column> to_durations(
  *
  * @throw cudf::logic_error if `durations` column parameter is not a duration type.
  *
- * @param durations Duration values to convert.
+ * @param durations Duration values to convert
  * @param format The string specifying output format.
- *        Default format is ""%d days %H:%M:%S".
- * @param mr Device memory resource used to allocate the returned column's device memory.
- * @return New strings column with formatted durations.
+ *        Default format is ""%D days %H:%M:%S".
```
Comment (mythrocks) [inline cpp/include/cudf/strings/convert/convert_durations.hpp:122]: Where do we stand on periods at the end of doxygen parameter descriptions? Remove them, like on line#120?
```suggestion
 *        Default format is ""%D days %H:%M:%S"
```

File: cpp/include/cudf/strings/convert/convert_floats.hpp (hunk 4 above)
Comment (mythrocks) [reply cpp/include/cudf/strings/convert/convert_floats.hpp:51]: I was about to flag this one, too.

File: cpp/include/cudf/strings/convert/convert_lists.hpp [hunk 21]
```diff
@@ -50,17 +50,19 @@ namespace strings {
...
+ * @param na_rep Replacement string for null elements
+ * @param separators Strings to use for enclosing list components and separating elements
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New strings column
  */
 std::unique_ptr<column> format_list_column(
   lists_column_view const& input,
-  string_scalar const& na_rep           = string_scalar("NULL"),
+  string_scalar const& na_rep           = string_scalar(""),
```
Comment (mythrocks) [inline cpp/include/cudf/strings/convert/convert_lists.hpp:62]: This is a change in behaviour.  Was this change intentional?  On the face of it, this seems unrelated to the task at hand.

Comment (davidwendt) [reply cpp/include/cudf/strings/convert/convert_lists.hpp:62]: This was intentional. The 'NULL' string-scalar requires allocating device memory using the default stream which will fail the stream tests. An empty string does not require allocating and device memory. One of the many reasons this PR is marked as breaking.

File: cpp/include/cudf/strings/convert/convert_durations.hpp (hunk 20 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/convert/convert_durations.hpp:122]: https://github.com/rapidsai/cudf/pull/14247#discussion_r1349788189

File: cpp/include/cudf/strings/convert/convert_lists.hpp (hunk 21 above)
Comment (mythrocks) [reply cpp/include/cudf/strings/convert/convert_lists.hpp:62]: P.S. Should we be removing this default?  I see that the default parameters have been removed in other locations.

**Review by mythrocks — COMMENTED** [review]
LGTM! A couple of minor questions.

File: cpp/include/cudf/strings/convert/convert_lists.hpp (hunk 21 above)
Comment (mythrocks) [reply cpp/include/cudf/strings/convert/convert_lists.hpp:62]: Ah, thanks for clarifying. That makes sense.

Comment (davidwendt) [reply cpp/include/cudf/strings/convert/convert_lists.hpp:62]: I see no reason to remove the new default.

File: cpp/include/cudf/strings/convert/convert_durations.hpp (hunk 20 above)
Comment (mythrocks) [reply cpp/include/cudf/strings/convert/convert_durations.hpp:122]: Thank you for that. Also, I wasn't aware of "Strunk & White".

**Review by mythrocks — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]
